kernel-headers/rdma/rdma_user_rxe.h

```diff
@@ -158,8 +158,54 @@ struct rxe_recv_wqe {
 	struct rxe_dma_info	dma;
 };
 
+struct rxe_uverbs_wc {
+	/* keep these the same as ib_uverbs_wc */
+	__aligned_u64		wr_id;
+	__u32			status;
+	__u32			opcode;
+	__u32			vendor_err;
+	__u32			byte_len;
+	union {
+		__be32		imm_data;
+		__u32		invalidate_rkey;
+	} ex;
+	__u32			qp_num;
+	__u32			src_qp;
+	__u32			wc_flags;
+	__u16			pkey_index;
+	__u16			slid;
+	__u8			sl;
+	__u8			dlid_path_bits;
+	__u8			port_num;
+	__u8			reserved;
+
+	/* any extras go here */
+	__aligned_u64		timestamp;
+	__aligned_u64		realtime;
+};
+
+enum rxe_capabilities {
+	RXE_CAP_NONE		= 0,
+	RXE_CAP_CMD_EX		= 1ULL << 0,
+};
+
+struct rxe_alloc_context_cmd {
+	__aligned_u64		provider_cap;
+};
+
+struct rxe_alloc_context_resp {
+	__aligned_u64		driver_cap;
+};
+
+struct rxe_create_cq_cmd {
+	__u8 is_extended;	/* called from ibv_create_cq_ex */
```

Comment:
The call path can't matter, we have three call paths for most commands, the driver data must be uniform across all of them.

Here is where it needs to be feature based. In this case I suggest you pass into the kernel the sizeof(rxe_uverbs_wc) so that the kernel knows how to format the completion queue entries. 0 means sizeof (ib_uverbs_wc).

Reply:
The standard pattern is to add a 

 u64 comp_mask;
 u64 flags;

Then use the flags as you need. The comp_mask is 0 (and checked to be 0 in the kernel) for future use.

Reply:
It might be wise to take the timestamp part out of here, then this series requires no more work. Add timestamp to user/kernel as another series, it will be easir to follow the compatability logic